### PR TITLE
Dynamicemb prefetch integration

### PR DESCRIPTION
**Pull request by JacoCheung**
## Description
In this PR, dynamicemb prefetch is integrated to hstu example. The user interface is  gin config file. When DynamicEmbeddingArgs is used together with `TrainerArgs.pipeline_type = "prefetch"`, the prefetch is automatically enabled. 

- Added a HBM cache example in dynamicemb . need @jiashuy's review and update. It would be better to change it to a notebook with clear comments.
- ~~Fix prefetch bug in dynamicemb: Remove `select` and `select_index` usage and use torch `index/nonzero` instead~~
- Fix misc bugs in dynamicemb.( add stream wait; fused_params param name mismatch between fbgemm tbe vs dynamicemb...)

### CODE REVIEW

File: corelib/dynamicemb/dynamicemb/key_value_table.py [hunk 1]
```diff
@@ -278,20 +276,15 @@ def find_impl(
...
-        num_missing_0: torch.Tensor = torch.empty(1, dtype=torch.long, device=device)
-        num_missing_1: torch.Tensor = torch.empty(1, dtype=torch.long, device=device)
-        missing_keys: torch.Tensor = torch.empty_like(unique_keys)
-        missing_indices: torch.Tensor = torch.empty(
-            batch, dtype=torch.long, device=device
-        )
-        select(missing, unique_keys, missing_keys, num_missing_0)
-        select_index(missing, missing_indices, num_missing_1)
-
+        missing_keys = unique_keys[missing]
```
Comment (shijieliu) [inline corelib/dynamicemb/dynamicemb/key_value_table.py]: I am not sure if this affects performance. why select and select_index dont work here?

File: corelib/dynamicemb/dynamicemb/key_value_table.py [hunk 2]
```diff
@@ -693,25 +679,23 @@ def lookup(
...
+            storage.value_dim()
+        )  # value is generally composed of embedding and optimizer state
 
         # 1. find in cache
         num_missing, missing_keys, missing_indices = cache.find_embeddings(
             unique_keys, unique_embs
         )
-        h_num_keys_for_storage = num_missing.cpu().item()
-        if h_num_keys_for_storage == 0:
+        if num_missing.item() == 0:
```
Comment (shijieliu) [inline corelib/dynamicemb/dynamicemb/key_value_table.py]: num_missing is a cpu tensor? why .cpu() is removed compared with previous implementation?

File: corelib/dynamicemb/dynamicemb/key_value_table.py [hunk 3]
```diff
@@ -741,11 +719,12 @@ def lookup(
             unique_embs[missing_indices, :] = embs_for_storage[:, :]
 
             # ignore the cache missed and storage missed when enable prefetch
+            # junzhang: WHY RETURN HERE?
             return
         else:
```
Comment (shijieliu) [inline corelib/dynamicemb/dynamicemb/key_value_table.py]: we dont need this else if we do direct return ahead

File: corelib/dynamicemb/example/hbm_as_cache_example.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
+import warnings
```
Comment (shijieliu) [inline corelib/dynamicemb/example/hbm_as_cache_example.py]: @jiashuy can we integrate this file with existing example?

File: corelib/dynamicemb/dynamicemb/key_value_table.py (hunk 1 above)
Comment (jiashuy) [reply corelib/dynamicemb/dynamicemb/key_value_table.py]: select and select_index is based on cub, which may be faster.

File: corelib/dynamicemb/dynamicemb/key_value_table.py [hunk 5]
```diff
@@ -278,20 +276,15 @@ def find_impl(
...
+        missing_keys = unique_keys[missing]
+        missing_indices = torch.nonzero(missing, as_tuple=True)[
+            0
+        ]  # missing is a 1D tensor; thereby we need to use [0] to get the indices
         if self._record_cache_metrics:
             self._cache_metrics[0] = batch
             self._cache_metrics[1] = founds.sum().item()
 
-        return num_missing_0, missing_keys, missing_indices
+        return missing.sum(), missing_keys, missing_indices
```
Comment (jiashuy) [inline corelib/dynamicemb/dynamicemb/key_value_table.py]: Hi @JacoCheung , I notice you call `missing.sum()` here, this is the case which I want to avoid.
Because sum will make the API synchronized, I want to leave the sync outside the Table Operation/API/

File: corelib/dynamicemb/example/hbm_as_cache_example.py (hunk 4 above)
Comment (jiashuy) [reply corelib/dynamicemb/example/hbm_as_cache_example.py]: I think we can integrate it

File: examples/hstu/test_utils.py [hunk 6]
```diff
@@ -459,9 +459,11 @@ def create_model(
         config=hstu_config,
         dynamicemb_options_dict={
             "item": DynamicEmbTableOptions(
-                global_hbm_for_values=0,
+                global_hbm_for_values=1024 * 1024,  # 1M HBM (maybe cached)
```
Comment (jiashuy) [inline examples/hstu/test_utils.py:462]: Make it larger? e.g. 1GB

File: corelib/dynamicemb/dynamicemb/key_value_table.py (hunk 5 above)
Comment (JacoCheung) [reply corelib/dynamicemb/dynamicemb/key_value_table.py]: Why torch.sum() is synced? the output should be a tensor on device as well, no?

File: corelib/dynamicemb/dynamicemb/key_value_table.py (hunk 2 above)
Comment (JacoCheung) [reply corelib/dynamicemb/dynamicemb/key_value_table.py]: [tensor.item()](https://docs.pytorch.org/docs/stable/generated/torch.Tensor.item.html) will do d2h implicitly, and it returns a standard python numeric. See also [Tensor.tolist()](https://docs.pytorch.org/docs/stable/generated/torch.Tensor.tolist.html)

File: corelib/dynamicemb/dynamicemb/key_value_table.py (hunk 3 above)
Comment (JacoCheung) [reply corelib/dynamicemb/dynamicemb/key_value_table.py]: This behavior is changed in PR #179.

File: corelib/dynamicemb/dynamicemb/key_value_table.py (hunk 5 above)
Comment (jiashuy) [reply corelib/dynamicemb/dynamicemb/key_value_table.py]: Oh yes, you are right, so the problem comes to why using the new one? is there any strong reason to replace the old one.

File: examples/hstu/test_utils.py (hunk 6 above)
Comment (JacoCheung) [reply examples/hstu/test_utils.py:462]: Why choose a larger size considering the vocab size  is  1024* 1024,  dim = 128, opt = ADAM ; emb storage is 3 * 1024* 1024 * 128 * 4 = 1536 MB. cahce size = 1 M. If set 1GB, most of the emb will residu on GPU.

Do we have some guidance on how to properly set `global_hbm_for_values`. We'd better make it clear to users if present. Especially with the example or documentations. @jiashuy

File: corelib/dynamicemb/dynamicemb/key_value_table.py (hunk 1 above)
Comment (JacoCheung) [reply corelib/dynamicemb/dynamicemb/key_value_table.py]: Yes, in terms of performance (pytorch overhead), select and select_index is better. But in terms of maintenance burden and code readability or cleanliness, using torch is better.

Using `select, select_index` is error-prone because it returns garbage values, users need to remove the tail garbage by slicing [0, num_return.item()) after call of `select, select_index`. 

There is a [bug](https://github.com/NVIDIA/recsys-examples/blob/6a3aeab7f379ffd7e753143a5000d627f746cfe7/corelib/dynamicemb/dynamicemb/key_value_table.py#L803) that I spotted. 

I am not sure if the perf loss is not acceptable or not.

File: corelib/dynamicemb/dynamicemb/key_value_table.py (hunk 5 above)
Comment (JacoCheung) [reply corelib/dynamicemb/dynamicemb/key_value_table.py]: Just for better cleanliness and readability.

File: corelib/dynamicemb/example/hbm_as_cache_example.py (hunk 4 above)
Comment (jiashuy) [reply corelib/dynamicemb/example/hbm_as_cache_example.py]: Hi @JacoCheung ， can you remove this file int this PR? And I will update it into example.py in another PR directly.
